5.10 hdmi cec rework

popcornmix opened the pull request:
This switches our CEC support to the version @mripard has cleaned up for upstreaming.
In testing it worked for me using kms on Pi3 and Pi4 in a range of HDMI modes.

mripard committed:
Revert "vc4_hdmi: Move hdmi reset to bind"

This reverts commit fdb112d44c1b0659aec1ef841939dff0f0dee49e.

mripard committed:
Revert "drm/vc4: Reading the hotplug register is only valid if no GPIO defined"

This reverts commit 266fd375dab37b8421671f6315ed261e59076bfb.

mripard committed:
Revert "vc4: cec: Restore cec physical address on reconnect"

This reverts commit 250344082a3957825d425ded325547e72a34fd65.

mripard committed:
Revert "vc4_hdmi: Remove cec_available flag as always supported"

This reverts commit 855cef03d18c9bc023a2787f9feac5476dbb4179.

mripard committed:
Revert "vc4_hdmi: Adjust CEC ref clock based on its input clock"

This reverts commit 7269a05437b3157a842f57bcb665a0801225702b.

mripard committed:
Revert "vc4_hdmi: Make irq shared"

This reverts commit cd59e087d8bba4ba834565b4d48ba983143b5f27.

mripard committed:
Revert "vc4_hdmi_regs: Make interrupt mask variant specific"

This reverts commit 37c74a07cf22a7385c93170a723d22b95ff20a51.

mripard committed:
Revert "vc4_hdmi_regs: Add Intr2 register block"

This reverts commit c865bb1bb6b481acfa4157e4331db278a176f887.

mripard committed:
Revert "vc4_hdmi: Fix register offset when sending longer CEC messages"

This reverts commit 3042fe2d9556f571fe52aae9df1275d9f3aa8b59.

mripard committed:
Revert "vc4_hdmi: BCM2835 requires a fixed hsm clock for CEC to work"

This reverts commit 1cad24365e9f4325d68d4a16025f77abe3eda2f7.

mripard committed:
Revert "vc4_hdmi: Fix up CEC registers"

This reverts commit 411944140cb321bca7e45a83dffcd1050866662c.

popcornmix committed:
bcm2711: Disable bsc_intr and aon_intr by default and enable in overlay

Signed-off-by: Dom Cobley <popcornmix@gmail.com>

popcornmix commented:
@pelwell I've attempted to disable by default and enable in the overlay but it's not working.
I think target should reference `interrupt-controller` but that's not working.

6by9 commented on arch/arm64/configs/bcmrpi3_defconfig:
```diff
@@ -1216,6 +1216,7 @@ CONFIG_MAXIM_THERMOCOUPLE=m
 CONFIG_MAX31856=m
 CONFIG_PWM_BCM2835=m
 CONFIG_PWM_PCA9685=m
+CONFIG_BRCMSTB_L2_IRQ=y
```
Why enable it for Pi3 defconfig? That seems odd?

popcornmix replied:
Good point - I'll remove it.

popcornmix replied:
(not entirely sure of the use of bcmrpi_3_defconfig - we use bcm2711_defconfig for official 64-bits on Pi3 and Pi4).

pelwell replied:
We could get rid of it, except that people are using it and might want the slightly smaller kernel on Pi 3.

pelwell commented on arch/arm/boot/dts/overlays/vc4-kms-v3d-overlay.dts:
```diff
@@ -115,6 +115,20 @@
 		};
 	};
 
+	fragment@15 {
```
This is the wrong overlay - you want vc4-kms-v3d-pi4.

popcornmix replied:
Ah...

pelwell commented:
Does the overlay now successfully enable the interrupt controllers?

mripard reviewed (commented):
We should probably not merge 961cd2f98034, or introduce it back.

It was meant to ease the rebasing of my branch, but it still looks useful

popcornmix commented:
Added "drm/vc4: Reading the hotplug register is only valid if no GPIO defined" back.

Yes, overlay changes appear successful.
We mask CEC interrupt on firmware and interrupts work on arm when vc4 overlay is enabled.
We do not mask CEC interrupt on firmware when vc4 overlay is not enabled and firmware cec works.

mripard committed:
drm/vc4: hdmi: Limit the BCM2711 to the max without scrambling

Unlike the previous generations, the HSM clock limitation is way above
what we can reach without scrambling, so let's move the maximum
frequency we support to the maximum clock frequency without scrambling.

Reviewed-by: Dave Stevenson <dave.stevenson@raspberrypi.com>
Signed-off-by: Maxime Ripard <maxime@cerno.tech>

popcornmix committed:
Revert "drm/vc4: hdmi: Limit the BCM2711 to the max without scrambling"

This reverts commit f5f5e195eea4a11585b2cba32c0ad65d6f0ed20b.

mripard committed:
ARM: bcm: Select BRCMSTB_L2_IRQ for bcm2835

The BCM2711 has a number of instances of interrupt controllers handled
by the driver behind the BRCMSTB_L2_IRQ Kconfig option (irq-brcmstb-l2).

Let's select that driver as part of the ARCH_BCM2835 Kconfig option.

Signed-off-by: Maxime Ripard <maxime@cerno.tech>
Acked-by: Florian Fainelli <f.fainelli@gmail.com>

popcornmix committed:
drm/vc4: hdmi: Move hdmi reset to bind

The hdmi reset got moved to a later point in the commit 9045e91a476b
("drm/vc4: hdmi: Add reset callback").

However, the reset now occurs after vc4_hdmi_cec_init and so tramples
the setup of registers like HDMI_CEC_CNTRL_1

This only affects pi0-3 as on pi4 the cec registers are in a separate
block

Fixes: 9045e91a476b ("drm/vc4: hdmi: Add reset callback")
Reviewed-by: Dave Stevenson <dave.stevenson@raspberrypi.com>
Signed-off-by: Dom Cobley <popcornmix@gmail.com>
Signed-off-by: Maxime Ripard <maxime@cerno.tech>

popcornmix committed:
drm/vc4: hdmi: Fix register offset with longer CEC messages

The code prior to 311e305fdb4e ("drm/vc4: hdmi: Implement a register
layout abstraction") was relying on the fact that the register offset
was incremented by 4 for each readl call. That worked since the register
width is 4 bytes.

However, since that commit the HDMI_READ macro is now taking an enum,
and the offset doesn't increment by 4 but 1 now. Divide the index by 4
to fix this.

Fixes: 311e305fdb4e ("drm/vc4: hdmi: Implement a register layout abstraction")
Reviewed-by: Dave Stevenson <dave.stevenson@raspberrypi.com>
Signed-off-by: Dom Cobley <popcornmix@gmail.com>
Signed-off-by: Maxime Ripard <maxime@cerno.tech>

popcornmix committed:
drm/vc4: hdmi: Fix up CEC registers

The commit 311e305fdb4e ("drm/vc4: hdmi: Implement a register layout
abstraction") forgot one CEC register, and made a copy and paste mistake
for another one. Fix those mistakes.

Fixes: 311e305fdb4e ("drm/vc4: hdmi: Implement a register layout abstraction")
Reviewed-by: Dave Stevenson <dave.stevenson@raspberrypi.com>
Signed-off-by: Dom Cobley <popcornmix@gmail.com>
Signed-off-by: Maxime Ripard <maxime@cerno.tech>

popcornmix committed:
drm/vc4: hdmi: Restore cec physical address on reconnect

Currently we call cec_phys_addr_invalidate on a hotplug deassert.
That may be due to a TV power cycling, or an AVR being switched
on (and switching edid).

This makes CEC unusable since our controller wouldn't have a physical
address anymore.

Set it back up again on the hotplug assert.

Fixes: 15b4511a4af6 ("drm/vc4: add HDMI CEC support")
Signed-off-by: Dom Cobley <popcornmix@gmail.com>
Signed-off-by: Maxime Ripard <maxime@cerno.tech>
Acked-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

mripard committed:
drm/vc4: hdmi: Compute the CEC clock divider from the clock rate

The CEC clock divider needs to output a frequency of 40kHz from the HSM
rate on the BCM2835. The driver used to have a fixed frequency for it,
but that changed for the BCM2711 and we now need to compute it
dynamically to maintain the proper rate.

Fixes: cd4cb49dc5bb ("drm/vc4: hdmi: Adjust HSM clock rate depending on pixel rate")
Reviewed-by: Dave Stevenson <dave.stevenson@raspberrypi.com>
Signed-off-by: Maxime Ripard <maxime@cerno.tech>

mripard committed:
drm/vc4: hdmi: Update the CEC clock divider on HSM rate change

As part of the enable sequence we might change the HSM clock rate if the
pixel rate is different than the one we were already dealing with.

On the BCM2835 however, the CEC clock derives from the HSM clock so any
rate change will need to be reflected in the CEC clock divider to output
40kHz.

Fixes: cd4cb49dc5bb ("drm/vc4: hdmi: Adjust HSM clock rate depending on pixel rate")
Reviewed-by: Dave Stevenson <dave.stevenson@raspberrypi.com>
Signed-off-by: Maxime Ripard <maxime@cerno.tech>

mripard committed:
drm/vc4: hdmi: Introduce a CEC clock

While the BCM2835 had the CEC clock derived from the HSM clock, the
BCM2711 has a dedicated parent clock for it.

Let's introduce a separate clock for it so that we can handle both
cases.

Reviewed-by: Dave Stevenson <dave.stevenson@raspberrypi.com>
Signed-off-by: Maxime Ripard <maxime@cerno.tech>

mripard committed:
drm/vc4: hdmi: Split the interrupt handlers

The BCM2711 has two different interrupt sources to transmit and receive
CEC messages, provided through an external interrupt chip shared between
the two HDMI interrupt controllers.

The rest of the CEC controller is identical though so we need to change
a bit the code organisation to share the code as much as possible, yet
still allowing to register independant handlers.

Signed-off-by: Maxime Ripard <maxime@cerno.tech>
Reviewed-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

mripard committed:
drm/vc4: hdmi: Support BCM2711 CEC interrupt setup

The HDMI controller found in the BCM2711 has an external interrupt
controller for the CEC and hotplug interrupt shared between the two
instances.

Let's add a variant flag to register a single interrupt handler and
deals with the interrupt handler setup, or two interrupt handlers
relying on an external irqchip.

Signed-off-by: Maxime Ripard <maxime@cerno.tech>
Reviewed-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

popcornmix committed:
drm/vc4: hdmi: Remove cec_available flag

Now that our HDMI controller supports CEC for the BCM2711, let's remove
that flag.

Reviewed-by: Dave Stevenson <dave.stevenson@raspberrypi.com>
Signed-off-by: Dom Cobley <popcornmix@gmail.com>
Signed-off-by: Maxime Ripard <maxime@cerno.tech>

mripard committed:
drm/vc4: hdmi: Don't register the CEC adapter if there's no interrupts

We introduced the BCM2711 support to the vc4 HDMI controller with 5.10,
but this was lacking any of the interrupts of the CEC controller so we
have to deal with the backward compatibility.

Do so by simply ignoring the CEC setup if the DT doesn't have the
interrupts property.

Reviewed-by: Dave Stevenson <dave.stevenson@raspberrypi.com>
Signed-off-by: Maxime Ripard <maxime@cerno.tech>

mripard committed:
dt-binding: display: bcm2711-hdmi: Add CEC and hotplug interrupts

The CEC and hotplug interrupts were missing when that binding was
introduced, let's add them in now that we've figured out how it works.

Signed-off-by: Maxime Ripard <maxime@cerno.tech>
Acked-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

mripard committed:
ARM: dts: bcm2711: Add the BSC interrupt controller

The BSC controllers used for the HDMI DDC have an interrupt controller
shared between both instances. Let's add it to avoid polling.

Reviewed-by: Florian Fainelli <f.fainelli@gmail.com>
Signed-off-by: Maxime Ripard <maxime@cerno.tech>

mripard committed:
ARM: dts: bcm2711: Add the CEC interrupt controller

The CEC and hotplug interrupts go through an interrupt controller shared
between the two HDMI controllers.

Let's add that interrupt controller and the interrupts for both HDMI
controllers

Reviewed-by: Florian Fainelli <f.fainelli@gmail.com>
Signed-off-by: Maxime Ripard <maxime@cerno.tech>

popcornmix committed:
bcm2711: Remove old GIC interrupt

Now handled through aon_intr

Signed-off-by: Dom Cobley <popcornmix@gmail.com>

popcornmix commented:
Updated to v2 of patches. cec now works on second hdmi port. Removed WIP.

popcornmix commented:
@mripard I'm getting: 
```
[   12.795523] irq 26, desc: 5ebbbeb4, depth: 0, count: 0, unhandled: 0
[   12.795549] ->handle_irq():  d8b055ac, brcmstb_l2_intc_irq_handle+0x0/0x148
[   12.795563] ->irq_data.chip(): 5666be55, gic_data+0x0/0xac
[   12.795570] ->action(): e4d27d3a
[   12.795582] ->action->handler(): bd6303dc, bad_chained_irq+0x0/0x6c
[   12.795590]      IRQ_LEVEL set
[   12.795596]    IRQ_NOPROBE set
[   12.795602]  IRQ_NOREQUEST set
[   12.795608]   IRQ_NOTHREAD set
[   12.795615] unexpected IRQ trap at vector 1a
```
after updating this PR. Can you spot something I've got wrong?

popcornmix commented:
I think this is good. I believe it's been accepted upstream and testing on Pi3 and Pi4 with kms worked.

popcornmix commented:
@pelwell okay to merge?

6by9 commented:
I rebased my tree this morning (still 5.10.13) and I appear to not be able to enable i2c_vc or i2c_arm successfully
```
[    6.153644] i2c-bcm2835 fe804000.i2c: Could not request IRQ
```
This is the main DT change going, but I can't immediately see any issue with it.

I'm doing a quick roll-back to try and confirm. Whilst that's building I'll do an rpi-update to check 5.10.14 out.

6by9 commented:
Sorry, I should add that it was when vc4-kms-v3d was loaded as well.

6by9 commented:
5.10.14-v7l from rpi-update files also fail with ```Could not request IRQ```.
I get /dev/i2c-11 and /dev/i2c-12 for the HDMI devices, but not /dev/i2c-1.
Remove ```dtoverlay=vc4-kms-v3d``` from config.txt and /dev/i2c-1 returns.

6by9 commented:
98a26692449ad74be15a25c05bb5be5396aaea4f (with a couple of WIP patches on top) is good.
